### ISSUE TITLE
GetOrCreateExclusive() and GetOrCreateExclusiveAsync(): Exclusive versions of GetOrCreate() and GetOrCreateAsync()

### ISSUE THREAD

**Issue by rmja** [aspnet/Caching#240]
It would be nice with equivalents of `GetOrCreate()` and `GetOrCreateAsync()` where it is ensured that the factory is only invoked once for each cache miss.

This basically resolves https://github.com/aspnet/Caching/issues/218, but with alternate extension methods.

An implementation proposal:

```
public static TItem GetOrCreateExclusive<TItem>(this IMemoryCache cache, object key, Func<ICacheEntry, TItem> factory)
{
    object result;
    
    if (!cache.TryGetValue(key, out result))
    {
        SemaphoreSlim exclusiveLock;

        lock (cache)
        {
            exclusiveLock = cache.GetOrCreate("__exclusive__", entry => {
                entry.Priority = CacheItemPriority.NeverRemove;
                return new SemaphoreSlim(1);
            });
        }

        exclusiveLock.Wait();

        try
        {
            if (cache.TryGetValue(key, out result))
            {
                return (TItem)result;
            }

            var entry = cache.CreateEntry(key);
            result = factory(entry);
            entry.SetValue(result);
            entry.Dispose();
        }
        finally
        {
            exclusiveLock.Release();
        }
    }

    return (TItem)result;
}

public static async Task<TItem> GetOrCreateExclusiveAsync<TItem>(this IMemoryCache cache, object key, Func<ICacheEntry, Task<TItem>> factory)
{
    object result;

    if (!cache.TryGetValue(key, out result))
    {
        SemaphoreSlim exclusiveLock;

        lock (cache)
        {
            exclusiveLock = cache.GetOrCreate("__exclusive__", entry => {
                entry.Priority = CacheItemPriority.NeverRemove;
                return new SemaphoreSlim(1);
            });
        }

        exclusiveLock.Wait();

        try
        {
            if (cache.TryGetValue(key, out result))
            {
                return (TItem)result;
            }

            var entry = cache.CreateEntry(key);
            result = await factory(entry); //.ConfigureAwait(false);
            entry.SetValue(result);
            entry.Dispose();
        }
        finally
        {
            exclusiveLock.Release();
        }
    }

    return (TItem)result;
}
```

**Comment by rmja** [comment]
In the async version it should of cause be `await exclusiveLock.WaitAsync();` and not the sync equivalent.

**Comment by ttrider** [comment]
Looking at the MemoryCache.cs implementation, I wonder how bad would it be to use ConcurrentDictionary there instead of plain Dictionary. and expose the GetOrAdd method directly.

I know that the interface doesn't have it and I would really hate IMemoryCache2, but at this point, I'm seriously considering forking implementation of the MemoryCache nd making change myself...

**Comment by JunTaoLuo** [comment]
@ttrider We are looking into updating the implementation of `MemoryCache` to use `ConcurrentDictionary` in #242. Note that the `GetOrAdd` method does not guarantee, by itself, that the factory is only invoked once and the implementation will involve `Lazy<T>`.

**Comment by ttrider** [comment]
Hello John,

 

This is the great news! While we wait on your implementation, for a time being, we created the following poor-man wrapper implementation.

https://gist.github.com/ttrider/a5ae8fc86ccfe6a4243f4481a5858a80

 

 

I think I’ll rename my methods to match your signatures – it will make dropping and replacing them easier.

 

Thank You,

 

Vladimir

**Comment by glennc** [comment]
I don't think this can fit into 2.0 right now. It needs a fairly large amount of investigation and testing to make it work well.

**Comment by ohadschn** [comment]
It would be great if in the async version, expiry could be set as a function of the retrieved value (see https://github.com/aspnet/Caching/issues/338).

**Comment by amit-ezra** [comment]
@JunTaoLuo 
it looks like the ConcurrentDictionary is already implemented, BUT the GetOrAdd method of the memory cache still isn't thread safe, even when using Lazy. it still doesn't use the back line ConcurrentDicionary.GetOrAdd...
I think it's very misleading have a non-atomic (non thread safe) GetOrAdd
is there a direct issue just on the GetOrAdd?

**Comment by jelical** [comment]
2 proposed solutions:
Based on Lazy - https://gist.github.com/jelical/17a3e1ef03af2901f4eb6866291239c6
Based on interlocks - https://gist.github.com/jelical/c424003efe008435b49fd54a3d1bd5b0
Both solutions are tested under heavy load, working with relatively the same performance, supporting multiple IMemoryCache instances and guarantee exclusive fetch call.

**Comment by dengere** [comment]
@jelical could you please share that your test code?
